ApollosProject/apollos-prototype
# Adds OneSignal pushId reporting to Apollos App

## DESCRIPTION

### What does this PR do, or why is it needed?

This PR Hooks into the OneSignal API on the GraphQL server, adding a RockID to a user on OneSignal as soon as a user is signed in, and we have push ids. 

The code in this PR is a little more complex than I would like, but I have a reason! There are three scenarios we need to account for to enable the mutation can be called properly. 

1. The user is logged in when the app boots, the Apollo cache is restored BEFORE the pushId is fetched from OneSignal. 
2. The user is logged in when the app boots, the pushId is fetched from OneSignal BEFORE the Apollos cache (containing auth token) is restored. 
3. A user has logged in, we already have a pushId from when the app booted. 

To accomplish this, I had to put the pushId in the Apollo store, as well as hook into the initial cache restore so it doesn't matter if the cache is restored first or the pushId is returned first, the mutation that uploads the pushId is triggered either way. 

### What design trade-offs/decisions were made?

Chose to increase complexity of code in 

### How do I test this PR?
1. Boot the simulator (should work on either iOS or Android simulator) 
2. Login to your user account.
3. Visit the ALL USERS page in OneSignal
https://onesignal.com/apps/b6e75a77-003f-4466-95ca-82cc5cdc407b/players
4. Turn on the "external device ids" field in the table.
<img width="1414" alt="screen shot 2019-01-09 at 3 25 10 pm" src="https://user-images.githubusercontent.com/1637694/50930835-a15c3c00-1426-11e9-9d4e-26d74978663b.png">
5. You should see your RockID listed next to your simulator. 


### What automated tests did you write?
n/a

## TODO:

- [x] Closes #472  
- [ ] No new warnings in tests, in storybook, and in-app. - **I made a new warning pop up on first boot that I'm struggling to fix. I think it has something to do with defining a client query before the schema is built or something**

## REVIEW:

**Manual QA**


**Code Review: Questions to consider**


> The purpose of PR Review is to _improve the quality of the software._
